File: item.js
/*
    mc4, a web voxel building game
    Copyright (C) 2019 kholland4

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

var allItems = {};

function getItemDef(itemstring) {
  var res = allItems[itemstring];
  if(res == undefined) { return null; }
  return res;
}

class ItemBase {
  constructor(itemstring, props) {
    this.itemstring = itemstring;
    
    this.desc = "";
    
    this.stackable = true;
    this.maxStack = 64;
    
    this.isNode = false;
    this.isTool = false;
    this.toolWear = 0;
    
    this.iconFile = null;
    this.icon = null;
    
    this.inCreativeInventory = true;
    
    //wood, tree, stone, sand, sandstone, flora, leaves
    this.groups = {};
    this.toolGroups = {}; //node groups that a tool "helps" with
    
    Object.assign(this, props);
    
    if(this.isTool) {
      this.stackable = false;
    }
    if(!this.stackable) {
      this.maxStack = 1;
    }
    
    if(this.iconFile != null && this.icon == null) {
      var e = iconExists(this.iconFile);
      if(e != false) {
        this.icon = e;
      } else {
        this.icon = loadIcon(this.itemstring, this.iconFile);
      }
    }
  }
}

class Item extends ItemBase {
  constructor(itemstring, props) {
    super(itemstring, props);
  }
}

api.Item = Item;
api.registerItem = function(item) {
  //TODO: validation
  allItems[item.itemstring] = item;
};
api.getItemDef = getItemDef;
api.listAllItems = function() {
  var out = [];
  for(var itemstring in allItems) {
    out.push(itemstring);
  }
  return out;
};

File: server/item.cpp
/*
    mc4 server
    Copyright (C) 2021 kholland4

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Affero General Public License for more details.

    You should have received a copy of the GNU Affero General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

#include "item.h"
#include "log.h"

std::map<std::string, ItemDef*> all_item_defs;

void load_item_defs(boost::property_tree::ptree pt) {
  size_t count = 0;
  
  for(const auto& n : pt.get_child("itemDefs")) {
    std::string itemstring = n.first;
    
    ItemDef *def = new ItemDef(itemstring);
    def->stackable = n.second.get<bool>("stackable");
    def->max_stack = n.second.get<int>("maxStack");
    
    def->is_node = n.second.get<bool>("isNode");
    def->is_tool = n.second.get<bool>("isTool");
    if(n.second.get<std::string>("toolWear") != "null")
      def->tool_wear = n.second.get<int>("toolWear");
    else
      def->tool_wear = 0;
    def->fuel = n.second.get<int>("fuel");
    
    def->in_creative_inventory = n.second.get<bool>("inCreativeInventory");
    
    const auto& groups_list = n.second.get_child("groups");
    for(const auto& group : groups_list) {
      def->groups.insert(group.first);
    }
    
    const auto& tool_groups_list = n.second.get_child("toolGroups");
    for(const auto& group : tool_groups_list) {
      std::string group_name = group.first;
      int max_level = group.second.get<int>("maxlevel");
      std::vector<double> times;
      for(const auto& time_item : group.second.get_child("times")) {
        times.push_back(std::stod(time_item.second.data()));
      }
      
      def->tool_groups[group_name] = std::make_pair(times, max_level);
    }
    
    all_item_defs[itemstring] = def;
    
    count++;
  }
  
  log(LogSource::LOADER, LogLevel::INFO, std::string("Loaded " + std::to_string(count) + " item definitions"));
}

ItemDef get_item_def(std::string itemstring) {
  auto search = all_item_defs.find(itemstring);
  if(search != all_item_defs.end()) {
    return *(search->second);
  }
  
  return ItemDef();
}
